nexus-and-alo/nexus-strawman.md

```diff
@@ -0,0 +1,361 @@
+
+**Nexus Task Handler:** Source code which is designed to handle incoming Nexus Tasks. Nexus Task Handler is similar to a Workflow or Activity definition and is hosted by a Nexus Worker. 
+
+**Nexus Worker:** A special purpose Worker that pulls Nexus Tasks for its registered Nexus Instances from the Server and executes them using associated Nexus Task Handlers. 
+
+## User flows
+
+### Hosting a Nexus Service
+
+1. User instantiates Nexus Service Instance using provided `nexus` CLI commands (also possible via SDK or API) and receives an Instance ID. Nexus is now able to receive requests to its frontend endpoint (though they don't get processed until step 4).
+1. User defines a Nexus Task Handler that will handle the processing of Nexus Tasks for the instantiated Nexus Service Instance.
+1. User hosts their Nexus Task Handler using a Nexus Worker. They must specify the Nexus Service Instance ID when starting the Nexus Worker.
+
+	// TODO (Chad): show how registration would work for generic handler
+    ```go
+    func main() {
+        ctx := context.TODO()
+
+        // Create Nexus client
+        client, err := client.Dial(ctx, client.Options{Target: "nexus-backend.example.com"})
+        if err != nil {
+            log.Fatal(err)
+        }
+        defer client.Close()
+
+        // Build worker with Temporal pieces
+        temporalClient, err := temporalclient.Dial(temporalclient.Options{HostPort: "nexus-backend.example.com:7233"})
+        if err != nil {
+            log.Fatal(err)
+        }
+        defer temporalClient.Close()
+        builder, err := temporalnexus.NewWorkerBuilder(temporalnexus.WorkerBuilderOptions{Client: temporalClient})
+        if err != nil {
+            log.Fatal(err)
+        }
+        // Add workflow as ALO to Nexus
+        err = builder.AddWorkflowALOHandler("foo", temporalnexus.WorkflowALOHandlerOptions{
+            Workflow: FooWorkflow,
+            Options:  temporalclient.StartWorkflowOptions{TaskQueue: "my-task-queue"},
+        })
+        if err != nil {
+            log.Fatal(err)
+        }
+
+        // Worker definition...
+    }
+    ```
+1. Nexus is now able to process requests to its frontend endpoint.
+
+### Consuming a Nexus Service from Temporal
+
+1. User asks the Nexus Service author for the endpoint of the Nexus Service Instance they are trying to consume. For Temporal based Nexus Service Instances, the endpoint will not be HTTP or gRPC but rather a custom uri formulation.
+2. User registers provided endpoint with their SDK to avoid hardcoding the endpoint into all of their application logic. 
+3. User calls the external Nexus Service using a Workflow-specific Nexus API which is provided by Temporal.
+    ```go
+    // calling a Nexus from a Temporal Workflow
+    func MyWorkflow(ctx workflow.Context, name string) (string, error) {
+        client := temporalnexus.NewAloClient(temporalnexus.AloClientOptions{Endpoint: "temporal:72aA64A0jhOL99B2"})
```

Comment:
The mapping should be done outside of the workflow.
Temporal will have a map of `nexus service type:service instance` -> `nexus endpoint`.

This simplifies writing workflows since you don't need to inject configuration into them and hides the transport detail from the workflow.

Some other ideas that came up during our earlier discussions were to assign an "environment" (e.g. prod / dev) to each worker so workflows don't have to specify the instance name, it'll default to the current environment.

Reply:
The whole idea of registration on the server side should be expanded in this document IMHO.

Reply:
Nexus service instance is a URI? So mapping is like `processOrder:mynexus.myco.com` -> `temporal:72aA64A0jhOL99B2`? And then you have no client and do this?

```go
alo, err := StartAlo(ctx, nexus.StartAloOptions{
  Type: "processOrder",
  Nexus: "mynexus.myco.com",
  ID: "my-business-id"
}
```

Reply:
+1 to removing the client layer.